lib/widgets/compose_box.dart

```diff
@@ -670,41 +680,140 @@ class _StreamContentInputState extends State<_StreamContentInput> {
   }
 }
 
-class _TopicInput extends StatelessWidget {
+class _TopicInput extends StatefulWidget {
   const _TopicInput({required this.streamId, required this.controller});
 
   final int streamId;
   final StreamComposeBoxController controller;
 
+  @override
+  State<_TopicInput> createState() => _TopicInputState();
+}
+
+class _TopicInputState extends State<_TopicInput> {
+  void _topicFocusChanged() {
+    setState(() {
+      if (widget.controller.topicFocusNode.hasFocus) {
+        widget.controller.topicInteractionStatus.value =
+          ComposeTopicInteractionStatus.isEditing;
+      } else if (!widget.controller.contentFocusNode.hasFocus) {
+        widget.controller.topicInteractionStatus.value =
+          ComposeTopicInteractionStatus.notEditingNotChosen;
+      }
```

Comment:
```suggestion
      if (widget.controller.topicFocusNode.hasFocus) {
        widget.controller.topicInteractionStatus.value =
          ComposeTopicInteractionStatus.isEditing;
      } else if (!widget.controller.contentFocusNode.hasFocus) {
        widget.controller.topicInteractionStatus.value =
          ComposeTopicInteractionStatus.notEditingNotChosen;
      } else {
        assert(widget.controller.topicInteractionStatus.value
          == ComposeTopicInteractionStatus.hasChosen);
      }
```

Does that look right?

Otherwise, it's a bit confusing because this appears to be missing one of the transitions in the state diagram. IIUC the point is that if the content input already has focus, then _contentFocusChanged should have already set the status.

Reply:
This `else` does represent the case when the focus moves from topic input to content input, however, the assertion will only succeed if `_contentFocusChanged` is called before `_topicFocusChanged`. Otherwise, the state is still `isEditing`; this seems to be the case consistently when I run all the tests.

> IIUC the point is that if the content input already has focus, then _contentFocusChanged should have already set the status.

Yeah, it appears that we have the up-to-date value of the content focus node (suggesting that `_topicFocusChanged` is called in the same frame?), but `_contentFocusChanged` hasn't been called yet. 

From the stacktrace:
```
#0      debugPrintStack (package:flutter/src/foundation/assertions.dart:1224:29)
#1      _TopicInputState._topicFocusChanged.<anonymous closure> (package:zulip/widgets/compose_box.dart:695:11)
#2      State.setState (package:flutter/src/widgets/framework.dart:1199:30)
#3      _TopicInputState._topicFocusChanged (package:zulip/widgets/compose_box.dart:685:5)
#4      ChangeNotifier.notifyListeners (package:flutter/src/foundation/change_notifier.dart:435:24)
#5      FocusNode._notify (package:flutter/src/widgets/focus_manager.dart:1131:5)
#6      FocusManager.applyFocusChangesIfNeeded (package:flutter/src/widgets/focus_manager.dart:1996:12)
```

`applyFocusChangesIfNeeded`:

```dart
    if (previousFocus != _primaryFocus) {
      assert(_focusDebug(() => 'Updating focus from $previousFocus to $_primaryFocus'));
      if (previousFocus != null) {
        _dirtyNodes.add(previousFocus);
      }
      if (_primaryFocus != null) {
        _dirtyNodes.add(_primaryFocus!);
      }
    }
    for (final FocusNode node in _dirtyNodes) {
      node._notify();
    }
```

I guess because the topic focus node is the `previousFocus`, and content focus node is the `_primaryFocus`, `_topicFocusChanged` always gets called before `_contentFocusChanged` when the focus moves from topic to content.

Reply:
It should be fine to set it to `hasChosen` twice; once in the else branch, another time in `_contentFocusChanged`.

However, `_contentFocusChanged` still needs to be kept. It handles the case when the focus moves from nothing to content input.